pkg/execution/state/state.go

```diff
@@ -104,6 +105,50 @@ type CustomConcurrency struct {
 	Limit int `json:"l"`
 }
 
+func (c CustomConcurrency) Validate() error {
+	// Keys must always be in the format of "$prefix:$id:$key", in which prefix
+	// is one of "f" | "e" | "a", depending on function, env, or account-level scopes.
+	if len(c.Key) < 5 {
+		return fmt.Errorf("invalid custom concurrency key length")
+	}
+	switch c.Key[0] {
+	case 'f', 'e', 'a':
+	default:
+		return fmt.Errorf("unknown concurrency key scope: %v", c.Key[0])
+	}
+	return nil
+}
+
+func (c CustomConcurrency) ParseKey() (scope enums.ConcurrencyScope, id uuid.UUID, err error) {
+	// Keys must always be in the format of "$prefix:$id:$key", in which prefix
+	// is one of "f" | "e" | "a", depending on function, env, or account-level scopes.
+	//
+	// An example key is `f:${uuid}:${hash}`.
+	if len(c.Key) < 5 {
+		return enums.ConcurrencyScopeFn, id, err
+	}
+
+	// TODO: Dont allocate, get index of colons and use offsets
+	parts := strings.Split(c.Key, ":")
+	if len(parts) != 3 {
+		// Invalid by default
+		return enums.ConcurrencyScopeFn, id, err
+	}
+
+	id, err = uuid.Parse(parts[1])
+
+	switch parts[0] {
+	case "f":
+		return enums.ConcurrencyScopeFn, id, err
```

Comment:
would like to see concurrency scope ints be explicitly defined, and also we can remove `gqlgen`

Reply:
✅ 9ff75544f71439c92e04d11d11965a4132a49c03